SSR: Whitelist (and validate) redirect_to for the /log-in page SSR

a8dar opened the pull request:
Before this commit, accessing a page like https://wordpress.com/log-in?redirect_to=https://wordpress.com/themes would have not been Server-Side-Rendered (only signup_flow & client_id were considered valid URL params for this). This is an issue because we are interested in generating a valid canonical URL for the pages that are indexed in search engines (Google, Bing, Yandex), and some of these search engines require a server-side-rendered HTML.

One way to easily enable SSR (which is currently strongly linked to the cache), is to add `redirect_to` to the `context.cacheQueryKeys` array. As we don't want to fill the cache with absolutely everything that may come through redirect_to, an additional `context.cacheQueryCallableValidators` was added which represents a hashmap (key to callable) for validating different URL (parameters. In case of redirect_to, only the URLs starting with https://wordpress.com/theme and https://wordpress.com/go would be cached (these are also the URLs that are currently reported as invalid by our SEO audit).

Semi-related note: an interesting aspect that I noticed is that `client_id` actually requires to have a `redirect_to` defined (otherwise the flow throws an exception). But, once redirect_to AND client_id are defined (which is always true when client_id is present), then the pages are not SSRed (because redirect_to was not on the list of whitelisted parameters). After this commit, nothing will change in this behavior, because only redirect_to that start with https://wordpress.com would be SSRed.

Testing:
1) Checkout the branxh, start npm with SSR debugging: `DEBUG='calypso:server-render' npm start`
2) Access http://calypso.localhost:3000/log-in - you should see the page as SSRed, with valid canonical and title (Log-In). Refresh the page and expect the same.
3) Access http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fgo%2Fweb-design%2Fleave-it-to-the-pros-outsource-your-website-security%2F and look at the Calypso debug logs. You should expect to see some SSR cache hit infos, but nothing that would say that URL is not cacheable.
4) Modify something in the URL above such that it doesn't start with wordpress.com (i.e. replace wordpress.com with bluehost). You should see in the logs a message saying that 'URL not cacheable because of failing the contraints for key redirect_to'.
5) Add one parameter to the URL in step 3. You should see that 'SSR: URL is not cacheable because it contains additional keys than the whitelist'
6) Repeat 3, 4 and 5 with two more URLs (one of them starting with themes also). Example: 1) https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Ftags%2F & 2) https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fthemes%2Ffilter%2Ffeature%3Avideo%2F
7) Look at the code and think of ways of breaking it.

jaswrks commented:
> Access http://calypso.localhost:3000/log-in - you should see the page as SSRed, with valid canonical and title (Log-In). Refresh the page and expect the same.

This looks good to me...

<img width="693" alt="2018-06-19_09-35-11" src="https://user-images.githubusercontent.com/1563559/41614643-7c473662-73a5-11e8-8a3f-ebefe2274674.png">

> Access http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fgo%2Fweb-design%2Fleave-it-to-the-pros-outsource-your-website-security%2F and look at the Calypso debug logs. You should expect to see some SSR cache hit infos, but nothing that would say that URL is not cacheable.

This looks right to me, based on your description.

<img width="728" alt="2018-06-19_09-38-12" src="https://user-images.githubusercontent.com/1563559/41614644-7c87ca4c-73a5-11e8-97ca-63e91c27ee42.png">

> Modify something in the URL above such that it doesn't start with wordpress.com (i.e. replace wordpress.com with bluehost). You should see in the logs a message saying something like 'URL not cacheable because of failing the contraints for key redirect_to'.

This looks right to me, based on your description.

<img width="724" alt="2018-06-19_09-39-37" src="https://user-images.githubusercontent.com/1563559/41614645-7cbc69aa-73a5-11e8-8cb1-b3861f5cb8f2.png">

> Add one parameter to the URL in step 3. You should see that in the debug logs that SSR is not enabled due to query constraints.

This looks right to me, based on your description.

<img width="631" alt="2018-06-19_09-40-56" src="https://user-images.githubusercontent.com/1563559/41614648-7cf2a998-73a5-11e8-987d-652ecb963c76.png">

> Repeat 3, 4 and 5 with two more URLs (one of them starting with themes also). Example: 1) https://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Ftags%2F & 2) https://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fthemes%2Ffilter%2Ffeature%3Avideo%2F

This looks right to me, based on your description.

<img width="730" alt="2018-06-19_09-43-09" src="https://user-images.githubusercontent.com/1563559/41614649-7d33dd6e-73a5-11e8-9ff2-9f0ab0c11521.png">

The one with `themes` in it is allowed.

<img width="823" alt="2018-06-19_09-44-06" src="https://user-images.githubusercontent.com/1563559/41614652-7d8c2212-73a5-11e8-9674-7663038e3457.png">

jaswrks commented on client/login/controller.js:
```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
+				return true; // nothing to check
+			}
+
+			const redirectToDecoded = decodeURI( context.query.redirect_to );
```
Did you already consider using `decodeURIComponent()` here?
I think it might be the better choice, but maybe there's a reason you chose not to use it.
Ref: https://stackoverflow.com/questions/747641/what-is-the-difference-between-decodeuricomponent-and-decodeuri

a8dar commented:
@jaswrks Thank you very much for the extensive test and the very professional way in which you took all the screenshots. Gives me a lot of confidence.

I have modified `decodeURI` to `decodeURIComponent`, since I was planning to use decodeURIComponent but autocomplete made it decodeURI.

jaswrks commented on client/login/controller.js:
```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
+			const redirectToDecoded = decodeURIComponent( context.query.redirect_to );
+			return (
+				redirectToDecoded.length < 120 &&
+				( redirectToDecoded.startsWith( 'https://wordpress.com/theme' ) ||
```
It might be worth making a mental note of the URLs listed here. I think vertical LPs are coming soon that will include `https://wordpress.com/websites/` and that might be another we should add to the whitelist.

sirreal commented on server/render/index.js:
```diff
@@ -226,3 +199,52 @@ export function serverRenderError( err, req, res, next ) {
+			debug( 'SSR Disabled due to the callable check on ' + validatorKey );
+			return false;
+		}
+	}
```
It would be nice to throw in development, but provide some safety in production with a function check

```js
if ( 'function' !== typeof validatorFunction ) {
  if ( 'production' === process.env.NODE_ENV ) {
    throw new TypeError( 'Terrible things happen!' );
  }
  return false;
}
```

ockham commented on client/login/controller.js:
```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
+export function setShouldSSRLogin( context, next ) {
+	if ( ! context.SSRValidators ) {
+		next(); // practically somehow the initSSRValidators was not initialized, so adding more is not necessary
+		return;
```
I've been using `return next()` in cases like this.

a8dar replied:
Not needed anymore, but I'll use return next() when I reach into a similar case.

ockham commented on client/login/controller.js:
```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
+				return true; // nothing to check
+			}
+
+			const redirectToDecoded = decodeURIComponent( context.query.redirect_to );
```
> Is any processing necessary here? I think page.js takes care of it.

We need to verify `Express` does it on the server side too, tho.

a8dar replied:
At the moment of the checks, redirect_to was still encoded, so I believe Express doesn't do it on the server side.

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,52 @@ export function serverRenderError( err, req, res, next ) {
+		locale: () => isDefaultLocale( context.lang ),
+		layout: () => context.layout,
+		query: () => Object.keys( context.query ).length === 0, // no query args
+	};
```
> Some of these conditions seem relevant for middlewares, some are app level. The app level ones might as well live in a centralized place and not let middlewares touch them.

This :+1:

ockham replied:
> … I wonder about this one. Any reason not to SSR other locales?
Can't do that right now, because `i18n-calypso` stores the current locale at module level. In a server setting, this would lead to cross-request pollution. Arguably something we should fix at some point, tho.

a8dar commented on client/login/controller.js:
```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
+
+			const redirectToDecoded = decodeURIComponent( context.query.redirect_to );
+			return (
+				redirectToDecoded.length < 120 &&
```
120: Just a value that is large enough to fit all URLs in the Google Doc where we have all the problematic canonicals. Some additional security to not start caching absolutely everything, but I removed it in the last commit, such that it doesn't create any confusion.

a8dar commented on server/render/index.js:
```diff
@@ -226,3 +199,52 @@ export function serverRenderError( err, req, res, next ) {
+ *
+ * Adds SSRValidators to the context, an object in which property names are idnetifiers and the values are callables
+ * that shouls return booleans. If one of the validators fails, then SSR is considered disabled. All methdos would
+ * receive the context on call.
```
Modified it to doSsr. I will also modify the function names to fit the convention.

sirreal commented on server/render/index.js:
```diff
@@ -145,26 +131,17 @@ export function serverRender( req, res ) {
-		! context.user &&
-		cacheKey &&
-		isDefaultLocale( context.lang )
+		isServerSideRenderCandidate( context ) &&
```
Should this function just return the result of all of the conditions here?

a8dar replied:
Depends, it's out choice how we group these conditions. If we decide to use only one function, then I wouldn't call it `isServerSideRenderCandidate`, but rather just `shouldServerSideRender`.

The current `isServerSideRenderCandidate` focuses on the app-checks directly related to the context (while doSSR would be more query related, and config is a global config). I guess the best solution (which I'll push) is to keep the current `isServerSideRenderCandidate` and add a new one `shouldServerSideRender` that includes all 3 conditions.

sirreal commented on client/login/controller.js:
```diff
@@ -159,3 +157,34 @@ export function redirectJetpack( context, next ) {
+		hasValidRedirectTo =
+			redirectToDecoded.startsWith( 'https://wordpress.com/theme' ) ||
+			redirectToDecoded.startsWith( 'https://wordpress.com/go' );
+	}
```
I understand that this list of valid `redirect_to` params is expected to grow quite large. Have you thought about how that should be handled? Maintaining a _very large list of `||` conditions_ doesn't appeal 😉 

---

I'd prefer avoiding the mutation. It also occurs to me that short-circuiting on `hasOnlyValidKeys` is nice and saves a potentially expensive set of checks, we may put `hasValidRedirectTo` into a state that's not accurate 🙂 

We might improve that if we pull `hasValidRedirectTo` out into a separate function, then we could keep the short circuiting and semantic correctness of `hasValidRedirectTo`:

```js
function isValidRedirectTo( redirectTo ) {
  if ( 'undefined' === typeof redirectTo ) {
    return true;
  }
  const decoded = decodeURIComponent( redirectTo );

  return decoded.startsWith( 'https://wordpress.com/theme' ) || …
}

export function setShouldServerSideRenderLogin( context, next ) {
  // …
  context.serverSideRender = hasOnlyValidKeys && isValidRedirectTo( context.query.redirect_to );
  next();
}
```

a8dar replied:
Like it, will modify it this way!
----
Regarding the ||, it feels keeping it this way for 2 values is the best variant due to the readability (i.e. if it starts with this or that, English-like). Once more values come in, depending on what those values really represent, this can evolve into something more complex (or can be combined with some Regexps).

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+ * @return {boolean} True if all the app-level criteria is fulfilled.
+ */
+function isServerSideRenderCompatible( context ) {
+	return Boolean(
```
Nit, I tend to use `!!` over `Boolean()`. (Do we even need that here?)

sirreal replied:
I'd say yes, it appears we need to coerce at least `context.layout` to boolean.

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+ * @param {function} next     As all middlewares, will call next in the sequence
+ */
+export function setShouldServerSideRender( context, next ) {
+	context.serverSideRender = Object.keys( context.query ).length === 0; // 0 query parameters expected
```
I think both the JSDoc and this comment could be a bit more explanatory:

* The JSDoc should maybe make clear that this is a middleware that's run early in the chain, and that sets a fallback for whether to SSR or not, which other middleware (run later in the chain) are allowed to override.
* The inline comment could explain the implementation detail that generally we don't want to SSR if there are any query args.

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+/**
+ * A generic middleware that identifies if the current request fits the conditions for being SSRable.
+ *
+ * Warning: Having serverSideRender=true is not sufficient. The app-level logic needs be applied before truly SSRing.
```
That reads a bit confusing to me, since it's not entirely clear whose responsibility it is to set what, and where. Furthermore, it might actually invite people to add section-specific logic here (which we want to avoid). (Also see my other comment, https://github.com/Automattic/wp-calypso/pull/25554/files#r198841859)

I think the JSDoc should make clear:
* This is a fallback middleware that sets `context.serverSideRender` early on
* If consumers need to override this at section level, they should set any section specific criteria by adding section specific middleware
* Even then, there are app level criteria that take precedence over `context.serverSideRender === true`

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+ *
+ * Warning: If this returns true, it is not sufficient for the page to be SSRed. Returning true from here is a
+ * pre-condition for SSR and the result needs to be corroborated with context.serverSideRender
+ * (context.serverSideRender is set to boolean by the middlewares, depending, in general, on the query params).
```
:+1:

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+ *
+ * Warning: if you think about calling this method or adding these conditions to the middlewares themselves (the ones
+ * that set context.serverSideRender), think twice: the context may not have all the necessary values;
+ * examples: context.layout, context.user may be set by other middlewares).
```
Maybe also add a note that section specific criteria should be handled by section level middleware, see https://github.com/Automattic/wp-calypso/pull/25554/files#r198843610

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+ *
+ * @param {object}   context The currently built context
+ *
+ * @return {boolean} True if all the app-level criteria is fulfilled.
```
`are` :slightly_smiling_face:

a8dar replied:
all is => all are.

ockham commented on server/render/index.js:
```diff
@@ -226,3 +199,57 @@ export function serverRenderError( err, req, res, next ) {
+}
+
+/**
+ * The main entry point for server-side rendering checks, and the final authority of if a page should be SSRed.
```
~`of`~

a8dar replied:
of is gone.

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,232 @@
+
+jest.mock( 'lib/i18n-utils', () => {
+	return {
+		isDefaultLocale: () => mockReturnValues.isDefaultLocale,
```
Hmm, why not just `=> undefined`? (Same for the instances below.) I find the detour through `mockReturnValues` a bit misleading here...

ockham replied:
Ohh, it took me a while to find `remock` https://github.com/Automattic/wp-calypso/pull/25554/files#diff-be817af57eab90eff3023a01926ac711R49

Hmm, still a bit confusing. Wonder if we have some established practices for patterns like that...

a8dar replied:
I have:
* moved `function remock` closer to `mockReturnValues`, just to express clearly the connection
* in the initial version I wanted to add const defaultMockContext = ... and have thew `mockReturnValues` assigned to this (to avoid the possible undefined confusion), but in the end I just put them in remock without creating a new const.
* added a DocBlock explaining a bit the behavior

I went in this direction just because by trying `jest.mock` inside a test didn't seem to work/override the global state. My intention is pretty simple: I just want to create a mock of those methods, and have them return a specific value for each test (preferably without writing tens of line of code in each of the tests)

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,232 @@
+	return fn;
+} );
+
+const aSsrCompatibleContext = {
```
I'd ditch the `a` article :smile: (partly because it sounds wrong, would need `an` before `S` since that's pronounced 'es'). Same below.

a8dar replied:
a is now gone. An alternative could have been `some`. My intention here was to clarify that it's not THE context, but rather some random context.

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,249 @@
+
+		setShouldServerSideRender( contextWithoutQueryKeys, next );
+		expect( contextWithoutQueryKeys.serverSideRender ).toBe( true );
+		expect( next.mock.calls.length ).toBe( 1 );
```
Can we use the more semantic `expect( next ).toHaveBeenCalledTimes( 1 );` here? See https://jest-bot.github.io/jest/docs/expect.html#tohavebeencalledtimesnumber

a8dar replied:
Nice syntax! All occurrences replaced.

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,249 @@
+
+		setShouldServerSideRender( contextWithQueryKeys, next );
+		expect( contextWithQueryKeys.serverSideRender ).toBe( false );
+		expect( next.mock.calls.length ).toBe( 1 );
```
`.toHaveBeenCalledTimes( 1 )`

ockham replied:
(And a few more instances below)

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,249 @@
+	} );
+} );
+
+// doing also the setShouldServerSideRenderLogin the test here because it seems adding it on the client side requires more effort
```
Hmm, having the test in a totally different directory than the thing it is testing is rather unfortunate. What would it take to move the test to `client/login/test/ssr.js` (where it should live)?

a8dar replied:
IIRC, I had the SSR functions inside the controller before. Running `npm test-client` did end up with plenty of errors when trying to import them from that location, so I had the impression I had to dig deeper into Jest to see what kind of bootstrapping is done before the test.

Since now the tests are in their own ssr.js, moving them to the right location worked like a charm.

ockham commented on server/render/test/index.test.js:
```diff
@@ -0,0 +1,249 @@
+/** @format */
```
The filename should be `index.js` to be aligned with Calypso nomenclature (we rely on the `test/` subdir to indicate that this is a test).

ockham commented:
(Oh, and let me know if you're busy with other things this week, I can make these changes myself if you prefer.)

a8dar commented:
@ockham I'll run a few more local tests and if it all passes, I'll merge.

Summary of most recent changes:
- toHaveBeenCalledTimes used everywhere (instead of mock.calls.length)
- renamed index.test.js to index.js according to the convention
- created the ssr.js test under client/controller/login and moved the login tests there.

ockham commented:
Thanks so much for your work on this @a8dar!
